### ISSUE TITLE
We should test the custom Vagrantfiles for every change through CI

### ISSUE THREAD

**Issue by LalatenduMohanty** [projectatomic/adb-atomic-developer-bundle#195]
We need to make sure https://github.com/projectatomic/adb-atomic-developer-bundle/blob/master/docs/using.rst#using-custom-vagrantfiles-for-specific-use-cases works always on libvirt as well on Virtualbox .

Because these are the the entry point of ADB experience for users. So the impact of these Vagrantfiles not working is huge on user experience. e.g. #194

**Comment by dharmit** [comment]
Hi! The way we're planning to work on this is - we will be testing a custom Vagrantfile on each merge to master when the Vagrantfile changes. This is to avoid testing when something other than the Vagrantfile changes (like README or shell scripts). Sounds good @praveenkumar @LalatenduMohanty ?

**Comment by LalatenduMohanty** [comment]
@dharmit if you mean CI will test the Vagrantfiles for each PR which is applicable to them then looks good to me.

**Comment by pradeepto** [comment]
@LalatenduMohanty  No, not every PR. But for every PR that is merged into the master repository. Doing the PR based CI needs much more work and without the plug-in place definitely requires more work. Also, PR based checks are generally used for static checks like Pylint, Unit tests etc (not saying that we can't use it for this case though).

I recommend that we first achieve CI for every merge first then take the next step.

**Comment by LalatenduMohanty** [comment]
@pradeepto I understand. I think we can work with CI for very PR that is merged into the master repository as you said (as ADB is a small project and we don't expect too many changes to the Vagrantfile) . 

However we need to keep in mind once the PR is merged in to master and has a bug in it , it becomes a technical debt for the project.  Where as when we detect the bug before it is merged the technical debt is only on the person who has sent the PR.

**Comment by pradeepto** [comment]
@LalatenduMohanty No. The word is continuous. The whole point is if after a merge if the CI fails it needs to be fixed and not let go and become a technical debt.

PR piplines are generally tertiary pipelines, at least in the ones I have worked in the past. I am all for them. And not just that, more the project advances we will need more pipelines in place - integration tests, nightly for long running tests if there are any,  acceptance tests et al. Obviously everything can't be achieved on day 1. So it is a continuous effort.

What I am trying to say is - let's achieve something first as a first goal. Then we can (and should) also have a PR pipeline. More so because we probably don't have the Github PR plugin at our disposal so we have to create the plumbing ourselves.

**Comment by dharmit** [comment]
What I intend to do after having played around with jenkins for some time is; use jenkins poll scm so that it polls the adb repository at fixed intervals. Whenever a new commit is detected, we execute a python script to check if there's change in any of the Vagrantfiles we are interested in. If there is a change, we test it using `vagrant up` and `kubectl`/`oc` commands. Sounds OK @praveenkumar @LalatenduMohanty ?

**Comment by pradeepto** [comment]
Just ot be clear, by Python script @dharmit  means that it would shell out to do some git operations to find out the diffs between the last known good SHA and the current HEAD.

**Comment by dharmit** [comment]
:+1:

**Comment by LalatenduMohanty** [comment]
@dharmit ok, lets get it working.

**Comment by praveenkumar** [comment]
@dharmit @pradeepto Ack, let work it out for first phase.

**Comment by dkutalek** [comment]
There is already relevant test written:
https://github.com/projectatomic/adb-tests/blob/master/testsuite/Components/vagrant/vagrantbox/sanity/runtest.sh

It is basic general sanity test to run every Vagrantfile provided, with no specifics as regard to Openshift etc. which differs across these Vagrantfiles.

Will discuss details with Dharmit today.

**Comment by dharmit** [comment]
We now have tests working upon PR. At the moment flow is like below:
- When a custom Vagrantfile for kubernetes or openshift is modified, a test is started on CentOS CI. This is not automated. One needs to comment `#build` on the PR. We can have behavior that everyone is OK with. :smile: 
- For other providers it's just that I've not written `test-<provider-name>.py` file yet. 
- Tests are executed using Python `unittest` module.
- To set things up, we're using Ansible.
- Code can be found under my forked repo - **[link](https://github.com/dharmit/adb-tests/tree/adb-fix-195-unittest/adb-vagrantfile-tests)**. It contains a `README` as well for reference.

@LalatenduMohanty @praveenkumar @navidshaikh thoughts?

**Comment by praveenkumar** [comment]
> One needs to comment #build on the PR. We can have behavior that everyone is OK with?

This `#build` is not auto provided by centos-ci user? If not then we have to document this in `how to contribute guide` so contributor should know how to test a PR.

> Tests are executed using Python unittest module.

+1, Something I can contribute to :+1: 

> To set things up, we're using Ansible.

where is Ansible script repo? 

> Code can be found under my forked repo - link. It contains a README as well for reference.

I want to contribute to openshift test cases if you can add me  as collaborator of that repo or should I go ahead and clone the test-repo (just don't want to duplicate same work which you did and not merged yet)?

**Comment by dharmit** [comment]
@praveenkumar 

> This #build is not auto provided by centos-ci user? If not then we have to document this in how to contribute guide so contributor should know how to test a PR.

`#build` is something you have to specify in the Jenkins job. You can have anything you want instead of `#build`. Another way is that every incoming PR is tested without anyone commenting `#build` on the PR. This was just to demonstrate that this works as well. The latter case of testing every PR automatically is definitely the most common. We will have to add details to contributing doc if we want to test PRs. I'll be adding that.

> where is Ansible script repo?

ATM, playbooks are located in [repo](https://github.com/dharmit/ci-ansible) under my gh namespace.

> I want to contribute to openshift test cases if you can add me as collaborator of that repo or should I go ahead and clone the test-repo (just don't want to duplicate same work which you did and not merged yet)?

Sure I can add you as collaborator. Haven't done that before so let's see how that goes. :smile:

**Comment by dkutalek** [comment]
Shanks has already written Ansible installers for RHEL and Fedora, for both ADB and CDK:
https://github.com/gsr-shanks/adb-deploy
https://github.com/gsr-shanks/cdk-install

Could you please discuss with Shanks possibilities of merging the efforts?

**Comment by dharmit** [comment]
@dkutalek Thanks for the links. I'll go through it and get in touch with Shanks.
